### PR TITLE
[SPARK-2883] [SQL] ORC data source for Spark SQL

### PR DESCRIPTION
**Pull request by liancheng**
This PR updates PR #6135 authored by @zhzhan from Hortonworks.

---

This PR implements a Spark SQL data source for accessing ORC files.

> **NOTE**
> 
> Although ORC is now an Apache TLP, the codebase is still tightly coupled with Hive.  That's why the new ORC data source is under `org.apache.spark.sql.hive` package, and must be used with `HiveContext`.  However, it doesn't require existing Hive installation to access ORC files.
## New Features
1.  Saving/loading ORC files without contacting Hive metastore
2.  Support for complex data types (i.e. array, map, and struct)
3.  Aware of common optimizations provided by Spark SQL:
   - Column pruning
   - Partitioning pruning
   - Filter push-down
## Future Work
1.  Schema evolution support
2.  Hive metastore table conversion
## Acknowledgements

This PR also include initial work done by @scwf from Huawei (PR #3753).

### CODE REVIEW

**Comment by liancheng** [comment]
@zhzhan Here is a rough list of my updates:

1.  Rebased to PR #6150, which updated the newly introduced partitioning support API

    Made corresponding changes according to the new API changes.

1.  `OrcFilters` update

    - I worked around the builder state inconsistency issue by employing a double-checking mechanism.  Now we can convert a single child of an `And` filter even if the other child is inconvertible.
    - Added data type checking, as ORC doesn't accept all Spark SQL atomic data types (e.g. timestamp).

1.  Added new tests

    Extracted some useful testing utility methods to `SQLTestUtils`, and added `NewOrcQuerySuite` based on `ParquetQuerySuite`.

1.  Some regular refactoring

    Mainly styling issues and code simplifications.

**Comment by liancheng** [comment]
Some TODO items related to testing:

- [ ] Cleanup current ORC test suites, as most of them are based on old Parquet test code, which has been deprecated and removed.
- [ ] More tests on filter push-down

**Comment by zhzhan** [comment]
@liancheng  Thanks for the followup. For the future work, feel free to assign to me.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+          rhs <- buildSearchArgument(right, lhs)
+        } yield rhs.end()
+
+        // For filter `left AND right`, we can still push down `left` even if `right` is not
+        // convertible, and vice versa.
+        conjunction
+          .orElse(tryLeft.flatMap(_ => buildSearchArgument(left, builder)))
+          .orElse(tryRight.flatMap(_ => buildSearchArgument(right, builder)))
+
+      case And(left, right) =>
```
Comment (zhzhan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala]: Should be Or?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    // For example, to convert an `And` filter with builder `b`, we call `b.startAnd()` first, and
+    // then try to convert its children.  Say we convert `left` child successfully, but find that
+    // `right` child is inconvertible.  Alas, `b.startAnd()` call can't be rolled back, and `b` is
+    // inconsistent now.
+    //
+    // The workaround employed here is that, for `And`/`Or`/`Not`, we first try to convert their
+    // children with brand new builders, and only do the actual conversion with the right builder
+    // instance when the children are proven to be convertible.
+    //
+    // P.S.: Hive seems to use `SearchArgument` together with `ExprNodeGenericFuncDesc` only.
```
Comment (zhzhan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala]: I checked with hive team. For external user, it is more expected to use the current builder  approach, although hive internally build xml file by ExprNodeGenericFuncDesc.

**Comment by zhzhan** [comment]
Jenkins, test this please.

**Comment by zhzhan** [comment]
@liancheng FYI: For schema merging, I checked some orc experts, and probably it is not supported the filter push down if the column is not in that specific orc file (I myself does not check the implementation yet). In the meantime, separating orc from hive is a on-going effort. We can separate orc from hive afterwards, and upgrade orc support to latest, which I think will improve the performance and a lot and remove potential version mismatch due to hive versions.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala (hunk 1 above)
Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala]: Oops, thanks!

**Comment by liancheng** [comment]
@zhzhan Thanks for the information.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala (hunk 2 above)
Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala]: Thanks. Do you know are there any other projects that uses ORC `SearchArgument` builder API? I'm looking for examples. I think the problem we faced should be pretty general. Would like to see how other projects solve it.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala [hunk 3]
```diff
@@ -143,6 +145,9 @@ private[sql] class SQLConf extends Serializable with CatalystConf {
   private[spark] def parquetUseDataSourceApi =
     getConf(PARQUET_USE_DATA_SOURCE_API, "true").toBoolean
 
+  private[spark] def orcFilterPushDown =
+    getConf(ORC_FILTER_PUSHDOWN_ENABLED, "false").toBoolean
```
Comment (scwf) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: why this is false by default?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/HadoopTypeConverter.scala [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        (value: Any, row: MutableRow, ordinal: Int) => row.setDouble(ordinal, oi.get(value))
+      case oi =>
+        (value: Any, row: MutableRow, ordinal: Int) => row(ordinal) = unwrap(value, oi)
+    }
+  }
+
+  /**
+   * Wraps with Hive types based on object inspector.
+   */
+  def wrappers(oi: ObjectInspector): Any => Any = wrapperFor(oi)
```
Comment (scwf) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/HadoopTypeConverter.scala]: how about add this two method to HiveInspectors instead of making a new object.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/package.scala [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.hive
+
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.sql.{DataFrame, SaveMode}
+
+package object orc {
```
Comment (rxin) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/package.scala]: why having this? this is not much more concise than 

sqlContext.load("orc", paths...)

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala (hunk 2 above)
Comment (scwf) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala]: is SearchArgument builder API stable/compatible for different hive version?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/package.scala (hunk 5 above)
Comment (zhzhan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/package.scala]: @rxin The code has been written for a while, before the new external data source API. Parquet and ORC are both very popular format. With this two API, user  operate on orc format quite similar to parquet format, which may be more convenient to end user.

Comment (rxin) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/package.scala]: What I'm saying is that in terms of total characters users have to type, it is actually strictly more to have this implicit conversion. Also implicit conversions tend to bring problems, impacting debuggability, binary compatibility, etc. We should just remove this.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/HadoopTypeConverter.scala (hunk 4 above)
Comment (zhzhan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/HadoopTypeConverter.scala]: @scwf good point. Actually after moving the first method to HiveInspectors, TableReader can use it as well. But doe it make sense to have a separate PR to do it, as this PR scope is quite limited to orc package, and has zero impact on any other features.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/package.scala [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+   * Extra ORC file loading functionality on [[HiveContext]] through implicit conversion.
+   *
+   * @since 1.4.0
+   */
+  @Experimental
+  implicit class OrcContext(sqlContext: HiveContext) {
+    /**
+     * ::Experimental::
+     *
+     * Loads specified Parquet files, returning the result as a [[DataFrame]].
```
Comment (tedyu) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/package.scala]: Parquet -> ORC

File: sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala (hunk 3 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: Filter push-down is a pretty error prone feature, 'cause it requires different components to interoperate seamlessly. We'd rather disable it by default until it's proven stable.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/HadoopTypeConverter.scala (hunk 4 above)
Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/HadoopTypeConverter.scala]: Thanks, moved them.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/package.scala (hunk 5 above)
Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/package.scala]: @rxin I removed them and added "orc" as a built-in data source. Also migrated other parts to the new DataFrame reader/writer API.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala (hunk 2 above)
Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala]: @scwf Good question. @zhzhan Would you mind helping confirming maturity of this API?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/package.scala (hunk 6 above)
Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/package.scala]: Oops... Thanks for pointing out the typo, but I just removed this file in favor of the new DataFrame reader/writer API :)

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala (hunk 2 above)
Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala]: @scwf BTW, with the help of the newly introduced isolated classloader mechanism, Spark SQL can always depend on the most recent version of Hive. At the meanwhile, users can specify arbitrary Hive metastore version to use. So even if this API changes across Hive versions, we don't need shim code to ensure compatibility.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala (hunk 3 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: How flaky is this? I'd maybe turn it on by default during QA, and then consider disabling it when we release if it is indeed flaky.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala (hunk 2 above)
Comment (scwf) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala]: get it, thanks for the explanation

**Comment by rxin** [comment]
LGTM with respect to API change (there isn't any).

**Comment by liancheng** [comment]
In the last a few commits, I added `"orc"` as a built-in data source name, so that we can have

```scala
hiveContext.read.format("orc").load("hdfs://...")
```

and

```scala
df.write.format("orc").save("hdfs://...")
```

Note that ORC data source is coupled with Hive. If users try to use it with `SQLContext`, an error message will be thrown to ask users to use `HiveContext` instead.

**Comment by liancheng** [comment]
@marmbrus This should be ready to go.

File: sql/core/src/main/scala/org/apache/spark/sql/sources/ddl.scala [hunk 7]
```diff
@@ -188,18 +188,20 @@ private[sql] class DDLParser(
 private[sql] object ResolvedDataSource {
 
   private val builtinSources = Map(
-    "jdbc" -> classOf[org.apache.spark.sql.jdbc.DefaultSource],
-    "json" -> classOf[org.apache.spark.sql.json.DefaultSource],
-    "parquet" -> classOf[org.apache.spark.sql.parquet.DefaultSource]
+    "jdbc" -> "org.apache.spark.sql.jdbc.DefaultSource",
+    "json" -> "org.apache.spark.sql.json.DefaultSource",
+    "parquet" -> "org.apache.spark.sql.parquet.DefaultSource",
+    "orc" -> "org.apache.spark.sql.hive.orc.DefaultSource"
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/ddl.scala]: Should we give a nice error here if not compiled with hive support?

**Comment by marmbrus** [comment]
Thanks guys!  Merged to master and 1.4.
